keps/sig-api-machinery/2558-field-version-openapi/README.md

```diff
@@ -0,0 +1,414 @@
+less common and less well-defined.
+
+3) Should we tag *all* fields, even those that are not gated, so that tooling
+can know that the absence of the tag is an error, rather than indicating GA.
+
+<<[/UNRESOLVED]>>
+
+### Example
+
+Consider adding a new alpha API field `Width` to the stable API `Frobber`
+in `v1.20`.
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // This field is alpha-level and is only honored by servers that enable the Frobber2D feature.
+  // +optional
+  Width  *int32 `json:"width,omitempty"
+}
+```
+
+The field `Width` is controlled by the `Frobber2D` feature gate defined
+in [`kube_features.go`].
+
+```go
+// owner: @you
+// alpha: v1.20
+//
+// Add multiple dimensions to frobbers.
+Frobber2D utilfeature.Feature = "Frobber2D"
+
+var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureSpec{
+  ...
+  Frobber2D: {Default: false, PreRelease: utilfeature.Alpha},
+}
+```
+
+To link the feature gate with the API field, the following comment line
+is added for the `Width` field:
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // +optional
+  // +fieldLifecycle=alpha
+  // +fieldMinVersion=v1.20
+  // +fieldFeatureGate=Frobber2D
+  Width  *int32 `json:"width,omitempty"
+}
+```
+
+This will generate an OpenAPI schema with an additional vendor extension field:
+
+```json
+"width": {
+  VendorExtensible: spec.VendorExtensible{
+    Extensions: spec.Extensions{
+      "x-kubernetes-field-version": []interface{}{
+        map[string]interface{}{
+          "fieldLifecycle":  "alpha",
+          "fieldMinVersion":  "v1.20",
+          "fieldFeatureGate": "Frobber2D",
+        },
+      },
+    },
+  },
+},
+```
+
+This shows that the field is alpha from v1.20 and 
+is controlled by the `Frobber2D` feature gate.
+
+### Constraints and Caveats
+
+#### Constraints
+
+The source of truth for versioning information for a feature gate
+will be the comment tags on the respective API fields because
+these comment tags are used for publishing to OpenAPI.
+
+A linter ensures that comments for feature gates defined in
+[`kube_features.go`] do not diverge from the comment lines on API fields.
+
+Additionally, the following constraints are applied and validated for
+comment tags on API fields:
+
+- If an API field contains the `fieldLifecycle` tag, it must contain
+the `fieldMinVersion` and `fieldFeatureGate` tags too.
+- The `fieldLifecycle` value must be one of `alpha`, `beta`, `stable`
+or `deprecated`.
+- The `fieldMinVersion` value must satisfy the regular expression `^v[1-9][0-9]*\.(0|[1-9][0-9]*)$`.
+- The `fieldFeatureGate` value is a valid feature gate as defined in
+[`kube_features.go`].
+
+### Test Plan
+
+The constraints for the comment tags listed above are validated using
+the [`hack/verify-description.sh`] script.
```

Comment:
I was thinking this could be a new [rule](https://github.com/kubernetes/kube-openapi/tree/master/pkg/generators/rules) in the [api_linter](https://github.com/kubernetes/kube-openapi/blob/master/pkg/generators/api_linter.go). I'm not sure how doable the `kube_features.go` lookup is though.